File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/RingBufferMath.java
package com.sedmelluq.discord.lavaplayer.tools;

import java.util.function.Function;

/**
 * Utility class for calculating averages on the last N values, with input and output transformers.
 */
public class RingBufferMath {
  private final double[] values;
  private final Function<Double, Double> inputProcessor;
  private final Function<Double, Double> outputProcessor;
  private double sum;
  private int position;
  private int size;

  /**
   * @param size Maximum number of values to remember.
   * @param inputProcessor Input transformer.
   * @param outputProcessor Output transformer.
   */
  public RingBufferMath(int size, Function<Double, Double> inputProcessor, Function<Double, Double> outputProcessor) {
    this.values = new double[size];
    this.inputProcessor = inputProcessor;
    this.outputProcessor = outputProcessor;
  }

  /**
   * @param value Original value to add (before transformation)
   */
  public void add(double value) {
    value = inputProcessor.apply(value);

    sum -= values[position];
    values[position] = value;
    sum += values[position];

    position = (position + 1) == values.length ? 0 : position + 1;
    size = Math.min(values.length, size + 1);
  }

  /**
   * @return Transformed mean of the internal values.
   */
  public double mean() {
    if (size == 0) {
      return outputProcessor.apply(0.0);
    } else {
      return outputProcessor.apply(sum / size);
    }
  }
}

File: main/src/main/java/lavaplayer/tools/RingBufferMath.kt
package lavaplayer.tools

import java.util.function.Function

/**
 * Utility class for calculating averages on the last N values, with input and output transformers.
 *
 * @param size            Maximum number of values to remember.
 * @param inputProcessor  Input transformer.
 * @param outputProcessor Output transformer.
 */
class RingBufferMath(
    private var size: Int,
    private val inputProcessor: Function<Double, Double>,
    private val outputProcessor: Function<Double, Double>
) {
    private val values = DoubleArray(size)
    private var sum = 0.0
    private var position = 0

    /**
     * @param value Original value to add (before transformation)
     */
    fun add(value: Double) {
        var value = value
        value = inputProcessor.apply(value)
        sum -= values[position]
        values[position] = value
        sum += values[position]
        position = if (position + 1 == values.size) 0 else position + 1
        size = values.size.coerceAtMost(size + 1)
    }

    /**
     * @return Transformed mean of the internal values.
     */
    fun mean(): Double {
        return outputProcessor.apply(if (size == 0) 0.0 else sum / size)
    }
}
